### ISSUE TITLE
Add support for -M/-MM

### ISSUE THREAD

**Issue by bc-lee** [ccache/ccache#1391]
I'm trying to build a project that uses `-MM` to generate dependencies. However, it seems that ccache does not support this option. I checked the ccache log and found the following:

```
Executing /usr/bin/gcc -Wno-maybe-uninitialized -Wshadow -O3 -ffast-math -m64 -Wall -I. -I. -std=gnu99 -D_GNU_SOUR      CE -mpreferred-stack-boundary=6 -fomit-frame-pointer -fno-tree-vectorize -fvisibility=hidden ./common/cpu.c -MT common/cpu.o -MM -g0
...
Compiler option -MM is unsupported
```

I'm not familiar with ccache's internals, but `-MMD` and `-MF` are already supported. It would be nice if ccache also supports `-MM`. For example, storing dependencies in a file and restoring them as stdout, just like the original compiler does. Thanks!

**Comment by jrosdahl** [comment]
Hi @bc-lee,

Note that `-MD`/`-MMD` are completely different: they produce dependency information _in addition to the object file from compilation_, while `-M`/`-MM` produce only dependency information. It's an alternative form of preprocessing, not compilation. See also [this Stack Overflow question and answer](https://stackoverflow.com/questions/29703938/ccache-doesnt-work-with-gcc-m-flag).

It would be technically possible to implement support for `-M`/`-MM`, but the gain would be small or negative. The reason why the gain would be small is that `-M`/`-MM` is already very fast compared to compilation. The gain would be negative if the direct mode cannot be used.

Here are some quick example benchmarks, compiling ccache's own `ccache.cpp` file on my system:

- Compilation (`-c`) with creating dependency file (`-MD`): 1780 ms
- The above with a ccache cache miss: 1886 ms
- The above with a ccache direct cache hit: 6 ms
- Preprocessing (`-E`): 67 ms
- Generating dependencies in Makefile form (`-M`): 38 ms

Implementing `-M` support the easy way and letting ccache use `-E` like it does for compilation would mean that a cache miss of a `-M` run would take in the order of 67 ms instead of 38 ms, thus making things slower. If we instead parse the output of `-M` then it would be possible to get down to much lower overhead.

In summary, I'm afraid it's most likely not worth the effort and maintenance cost to add support for `-M`/`-MM`.

**Comment by jrosdahl** [comment]
Closing due to reasons mentioned above.
